agent/utils/utils.go

```diff
@@ -327,6 +354,70 @@ type ResponseForwarder struct {
 	metricHandler *metrics.MetricHandler
 }
 
+func newBufferedReadSeeker(r io.Reader, bufSize int) *bufferedReadSeeker {
+	return &bufferedReadSeeker{
+		r:         r,
+		buf:       make([]byte, bufSize),
+		writeHead: 0,
+		readHead:  0,
+	}
+}
+
+type bufferedReadSeeker struct {
+	r         io.Reader
+	buf       []byte
+	writeHead int
+	readHead  int
+}
+
+func (b *bufferedReadSeeker) Read(p []byte) (int, error) {
+	// Read from buffer.
+	readFromBuf := copy(p, b.buf[b.readHead:b.writeHead])
+	b.readHead += readFromBuf
+	// Read from wrapped source and write to buffer.
+	readFromSource, err := b.r.Read(p[readFromBuf:])
+	written := copy(b.buf[b.writeHead:], p[readFromBuf:(readFromBuf+readFromSource)])
+	b.writeHead += written
+	b.readHead += written
+	return readFromBuf + readFromSource, err
+}
+
+func (b *bufferedReadSeeker) Seek(offset int64, whence int) (int64, error) {
+	if whence != io.SeekStart {
+		return 0, errors.New("only SeekStart offset is supported")
+	}
+	if offset < 0 || offset >= int64(len(b.buf)) {
+		return 0, errors.New("invalid offset value")
+	}
+	if b.writeHead >= len(b.buf) {
+		return 0, errors.New("cannot seek, possible buffer overflow")
+	}
+	b.readHead = int(offset)
+	return int64(b.readHead), nil
+}
+
+func postResponseWithRetries(client *http.Client, proxyReq *http.Request, proxyReadSeeker io.ReadSeeker) error {
```

Comment:
Similar to the `getRequestWithRetries` method, I think it would be better if this one constructed the request rather than having it passed in.

Reply:
Done.